paig-server/backend/paig/api/shield/services/application_manager_service.py

```diff
@@ -42,39 +44,47 @@ def load_scanners(self, application_key):
         self.application_key_scanners.put(application_key, scanner_list)
         logger.info(f"Found {scanner_list} scanners for application key: {application_key}")
 
-    def get_scanners(self, application_key):
+    def get_scanners(self, application_key, is_authz_scan) -> list:
         """
         Get the scanners for the given application key.
         If the scanners are not in the cache, load them.
 
         Args:
             application_key (str): The application key.
+            is_authz_scan (bool): The flag to determine if the scan is an authz or non authz.
 
         Returns:
             list: The list of scanners for the application key.
         """
         if application_key not in self.application_key_scanners.cache:
             self.load_scanners(application_key)
-        return self.application_key_scanners.get(application_key)
 
-    def scan_messages(self, application_key, message):
+        all_scanners = self.application_key_scanners.get(application_key)
+
+        scanners_list = [
+            scanner for scanner in all_scanners
+            if getattr(scanner, 'enforce_access_control', False) == is_authz_scan
+        ]
+
+        return scanners_list
+
+    def scan_messages(self, application_key, message, is_authz_scan) -> (dict[str, ScannerResult], dict[str, str]):
         """
         Scan the given messages for all the scanners where the enforce access control flag is true.
 
         Args:
             application_key (str): The application key.
             message (str): The message to scan.
+            is_authz_scan (bool): The flag to determine if the scan is an authz or non authz.
 
         Returns:
             tuple: A tuple containing the scan results and the access control results.
         """
 
-        scanners = self.get_scanners(application_key)
+        scanners = self.get_scanners(application_key, is_authz_scan)
         logger.debug(f"Found {len(scanners)} scanners for application key: {application_key}")
 
-        scan_results = {}
-        scan_timings = {}
-        access_control_traits = []
+        scan_results, scan_timings = {}, {}
         with concurrent.futures.ThreadPoolExecutor(max_workers=self.shield_scanner_max_workers) as executor:
             future_to_scanner = {executor.submit(scan_with_scanner, scanner, message): scanner for scanner in scanners}
```

Comment:
We should filter the list of scanners that are applicable at this point of the conversation - not all scanners will be application to all points of conversation - prompt, enriched-prompt, rag, response

Reply:
Added logic for the same